### PR TITLE
feat(admin): add custom admin route ranking feature

### PR DESCRIPTION
**Pull request by bqst**
## Summary

**What** — What changes are introduced in this PR?

CleanShot 2025-11-03 at 17 38 52

This PR adds a `rank` property to the `RouteConfig` interface, allowing developers to control the order of custom admin routes in the sidebar menu. Routes are sorted in ascending order by rank value (lower rank appears first), with unranked routes appearing after all ranked routes.

**Why** — Why are these changes relevant or necessary?  

When we start creating multiple custom admin UI routes, they need a way to organize and order these routes in a predictable manner. Without ranking, routes appear in an arbitrary order, making it difficult to create a logical, user-friendly navigation structure. This feature follows the same pattern used by Medusa's `product_category.rank` field, providing consistency across the framework.

**How** — How have these changes been implemented?

1. **Type Definitions** - Added `rank?: number` property to `RouteConfig` interface in `@medusajs/admin-sdk`
2. **Config Parsing** - Updated the vite plugin's `generate-menu-items.ts` to extract and include rank values from route configs
3. **Sorting Logic** - Implemented `sortMenuItemsByRank()` method in the dashboard app that:
   - Sorts routes by rank in ascending order
   - Places routes with rank before routes without rank
   - Recursively sorts nested routes independently
   - Maintains stable sort for routes with equal or undefined ranks
4. **Documentation** - Updated `learn/fundamentals/admin/ui-routes` with comprehensive rank usage examples
5. **Tests** - Added 14 new tests covering menu item generation and sorting logic

**Testing** — How have these changes been tested, or how can the reviewer test the feature?

- ✅ **14 automated tests added** - All passing
  - 4 tests for menu item generation with rank values
  - 10 tests for sorting logic (including edge cases)
- ✅ **Manual testing** in test-project with multiple routes demonstrating:
  - Top-level route ranking
  - Nested route ranking
  - Mixed ranked and unranked routes
  - Routes nested under existing admin routes (Orders)
- ✅ **Verified locally** that routes appear in correct order based on rank

---

### Basic Usage - Top Level Routes

```tsx
// src/admin/routes/analytics/page.tsx
import { defineRouteConfig } from "@medusajs/admin-sdk"
import { ChartBar } from "@medusajs/icons"

export const config = defineRouteConfig({
  label: "Analytics",
  icon: ChartBar,
  rank: 1, // Will appear first
})
```

```tsx
// src/admin/routes/reports/page.tsx
import { defineRouteConfig } from "@medusajs/admin-sdk"
import { DocumentText } from "@medusajs/icons"

export const config = defineRouteConfig({
  label: "Reports",
  icon: DocumentText,
  rank: 2, // Will appear second
})
```

Result: **Analytics** appears before **Reports** in the sidebar.

### Nested Routes with Ranking

```tsx
// src/admin/routes/custom/nested-1/page.tsx
export const config = defineRouteConfig({
  label: "First Nested",
  rank: 1, // Appears first under parent
})
```

```tsx
// src/admin/routes/custom/nested-2/page.tsx
export const config = defineRouteConfig({
  label: "Second Nested",
  rank: 2, // Appears second under parent
})
```

### Routes Nested Under Existing Admin Routes

```tsx
// src/admin/routes/orders/insights/page.tsx
export const config = defineRouteConfig({
  label: "Order Insights",
  nested: "/orders",
  rank: 1, // First under Orders menu
})
```

```tsx
// src/admin/routes/orders/reports/page.tsx
export const config = defineRouteConfig({
  label: "Order Reports",
  nested: "/orders",
  rank: 2, // Second under Orders menu
})
```

### Mixed Ranked and Unranked Routes

```tsx
// Routes without rank appear after all ranked routes
export const config = defineRouteConfig({
  label: "Unranked Page",
  // No rank - will appear last
})
```

---

## Checklist

Please ensure the following before requesting a review:

- [x] I have added a **changeset** for this PR
    - Every non-breaking change should be marked as a **patch**
    - To add a changeset, run `yarn changeset` and follow the prompts
- [x] The changes are covered by relevant **tests**
- [x] I have verified the code works as intended locally
- [x] I have linked the related issue(s) if applicable

---

### Why Ranking is Important

When building custom admin interfaces, developers often create multiple custom routes. Without a ranking system, these routes appear in unpredictable order, leading to:
- Poor user experience with illogical menu organization
- Difficulty presenting important routes first
- No control over nested route order within parent menus

This feature becomes essential as admin customizations grow. For example, an e-commerce store might want:
1. **Dashboard** (rank: 1) - Most important, shown first
2. **Analytics** (rank: 2) - Secondary priority
3. **Reports** (rank: 3) - Lower priority
4. **Settings** (no rank) - Shown last

### Implementation Notes

- **Pattern Consistency**: Follows the same rank pattern as `product_category.rank` in Medusa
- **Backward Compatible**: Routes without rank continue to work (appear after ranked routes)
- **Zero and Negative Ranks**: Supported for advanced use cases
- **Recursive Sorting**: Nested items are sorted independently from their parents
- **Stable Sort**: Routes with equal ranks maintain their original order

### CODE REVIEW

File: packages/admin/dashboard/src/dashboard-app/dashboard-app.tsx [hunk 1]
```diff
@@ -196,9 +197,50 @@ export class DashboardApp {
...
+    // Sort menu items by rank (ascending order, undefined ranks come last)
+    registry.forEach((items, key) => {
+      const sorted = this.sortMenuItemsByRank(items)
+      registry.set(key, sorted)
+    })
+
     return registry
   }
 
+  private sortMenuItemsByRank(
```
Comment (fPolic) [inline packages/admin/dashboard/src/dashboard-app/dashboard-app.tsx]: should we extract the sorting logic into a util function that is called here and in the test?

**Review by fPolic — COMMENTED** [review]
Nice contribution @bqst! Left one comment.

**Review by shahednasser — CHANGES_REQUESTED** [review]
Thank you for your contribution @bqst ! Can I get you to move the documentation contribution to a separate PR? That way it's only merged when the feature is released.

File: packages/admin/admin-vite-plugin/src/routes/generate-menu-items.ts [hunk 2]
```diff
@@ -240,10 +245,22 @@ function processConfigProperties(
     return null
   }
 
+  const rank = properties.find(
+    (prop) =>
+      isObjectProperty(prop) && isIdentifier(prop.key, { name: "rank" })
+  ) as ObjectProperty | undefined
+
+  let rankValue: number | undefined = undefined
```
Comment (willbouch) [inline packages/admin/admin-vite-plugin/src/routes/generate-menu-items.ts]: is = undefined necesary?

**Review by willbouch — COMMENTED** [review]
Shahed's comment to address and I will be ready to approve

File: packages/admin/admin-vite-plugin/src/routes/generate-menu-items.ts (hunk 2 above)
Comment (bqst) [reply packages/admin/admin-vite-plugin/src/routes/generate-menu-items.ts]: You're absolutely right!

**Review by shahednasser — APPROVED** [review]

**Review by fPolic — APPROVED** [review]

File: .changeset/metal-lamps-film.md [hunk 3]
```diff
@@ -0,0 +1,7 @@
+---
+"@medusajs/admin-vite-plugin": minor
+"@medusajs/admin-sdk": minor
+"@medusajs/dashboard": minor
```
Comment (fPolic) [inline .changeset/metal-lamps-film.md]: todo: also patch

File: .changeset/metal-lamps-film.md [hunk 4]
```diff
@@ -0,0 +1,7 @@
...
+---
+
+add custom admin route ranking feature
```
Comment (fPolic) [inline .changeset/metal-lamps-film.md]:
```suggestion
feat(dashboard): add custom admin route ranking feature
```

**Comment by willbouch** [comment]
@bqst added the automerge flag. So as soon as the conflicts are resolved and the tests pass, it will merge itself
